Combine some access to the thread local variable for script thread.

Narfinger opened the pull request:
This combines some access to the thread local variable for script
thread.

- We introduce a new UserInteractingScriptGuard which on drop handles
  the resetting of was_interacting to the previous value. Sometimes
throughout the code `ScriptThread::is_user_interacting` was reset to the
previous value while sometimes just set to false. This should
remove this footgun.
- This also reduces the amount of thread local access for MutationObservers and task queue.

Testing: WPT tests should cover this.
Fixes: This addresses part of https://github.com/servo/servo/issues/37969 but there is probably still stuff to be done.

jdm commented:
I think the two main changes you describe are promising, but I'm not yet convinced that trying to batch fetches if the ScriptThread TLS via with_script_thread is a good tradeoff. I'd like to experiment with the solution I was imagining (passing a &ScriptThread argument to more methods) to see if it's realistic.

Narfinger committed:
Reduce the amount of script thread thread local access.

This introduces two changed:
- We introduce a new UserInteractingScriptGuard which on drop handles
the resetting of was_interacting to the previous value. Sometimes
throughout the code ScriptThread::is_user_interacting was reset to the
previous value while sometimes just set to false. This should
remove this footgun.
- mutation_observers and mutation_observer_microtask_queued are now in a
  common structure with Rc's inside of them to easily get the new
structure when both have to be modified.

Signed-off-by: Narfinger <Narfinger@users.noreply.github.com>

Narfinger commented:
Ok I used your changes, the mutation observer changes and the UserInteractingGuard to make this PR now. Hope we can get this merged and I can work on more of it when I find more time.

jdm commented on components/script/script_thread.rs:
```diff
@@ -192,6 +192,43 @@ unsafe_no_jsmanaged_fields!(TaskQueue<MainThreadScriptMsg>);
+#[derive(Default)]
+pub(crate) struct ScriptUserInteractingGuard<'a> {
+    was_interacting: bool,
+    script_thread: Option<&'a ScriptThread>,
```
If we make `ScriptThread::is_user_interacting` a `Rc<Cell<bool>>`, we don't need to store a pointer to the ScriptThread. This allows us to remove all of the new `with_script_thread` calls that were added, which will make the changes much cleaner.

jdm commented on components/script/script_thread.rs:
```diff
@@ -477,33 +510,23 @@ impl ScriptThread {
-                .collect()
+                .mutation_observer_microtask_queued
+                .clone(),
+            mutation_observers: script_thread.mutation_observers.mutation_observers.clone(),
```
Cloning this list means we won't affect the original. We need to return a `Rc<MutationObservers>` from this method instead, and store that inside ScriptThread.

jdm commented on components/script/script_thread.rs:
```diff
@@ -192,6 +192,43 @@ unsafe_no_jsmanaged_fields!(TaskQueue<MainThreadScriptMsg>);
+#[cfg_attr(crown, allow(crown::unrooted_must_root))]
+pub(crate) struct MutationObservers {
+    /// Microtask Queue for adding support for mutation observer microtasks
+    pub(crate) mutation_observer_microtask_queued: Rc<Cell<bool>>,
```
If we're using `Rc` around this struct, we won't need the Rc on this field.

jdm reviewed (changes requested):
I think these changes will be more clearly beneficial after my suggestions! The other benefit to using Rc around MutationObservers is we can then store it inside Document, and that should make getting rid of `ScriptThread::mutation_observers` in the future an easier task.

jdm commented on components/script/dom/mutationobserver.rs:
```diff
@@ -93,33 +95,42 @@ impl MutationObserver {
 
     /// <https://dom.spec.whatwg.org/#notify-mutation-observers>
     pub(crate) fn notify_mutation_observers(can_gc: CanGc) {
+        let mutation_observers = ScriptThread::mutation_observers();
```
I would put it in the root directory of the script crate.

Narfinger commented on components/script/script_thread.rs:
```diff
@@ -192,6 +192,43 @@ unsafe_no_jsmanaged_fields!(TaskQueue<MainThreadScriptMsg>);
+
+/// A helper struct for mutation observers
+#[derive(JSTraceable, Default)]
+#[cfg_attr(crown, allow(crown::unrooted_must_root))]
```
Currently crown complains about the unrooted_must_root on the mutation_observers.

jdm replied:
You can use https://github.com/servo/servo/blob/ae2d9674e8c47d9df53fd11129e3f9f7dbfdc15b/components/script/dom/promise.rs#L47 in this case, since the Rc is always stored in ScriptThread, so it's always reachable by the GC.

Narfinger replied:
Still complaining.
```
   Compiling script v0.0.1 (/home/cengels/dev/servo/components/script)
error: Type must be rooted, use #[crown::unrooted_must_root_lint::must_root] on the struct definition to propagate.
  --> components/script/mutation_observers.rs:27:5
   |
27 |     pub(crate) mutation_observers: DomRefCell<Vec<Dom<MutationObserver>>>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(crown::unrooted_must_root)]` on by default

error: could not compile `script` (lib) due to 1 previous error```

jdm committed:
Fix crown rooting error.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

jdm committed:
Move all remaining mutation observer logic out of ScriptThread.

Signed-off-by: Josh Matthews <josh@joshmatthews.net>

jdm commented on components/script/script_thread.rs:
```diff
@@ -477,36 +498,30 @@ impl ScriptThread {
-    pub(crate) fn get_mutation_observers() -> Vec<DomRoot<MutationObserver>> {
-        with_script_thread(|script_thread| {
-            script_thread
+    pub(crate) fn mutation_observers() -> MutationObservers {
```
https://github.com/Narfinger/servo/pull/1

jdm commented:
I'm pretty sure these are legitimate correctness changes because of your user interaction changes. We can update the expect results and move on.

```
OK /webxr/xrDevice_requestSession_immersive_no_gesture.https.html

    FAIL [expected PASS] subtest: Requesting immersive session outside of a user gesture rejects

    assert_unreached: Should have rejected: undefined Reached unreachable code

OK /webxr/xrDevice_requestSession_requiredFeatures_unknown.https.html

    PASS [expected FAIL] subtest: Tests requestSession rejects for unknown requiredFeatures
``

Narfinger commented:
Ok let's try it now.
